# 01 — Inicijalni pregled skupa podataka

U ovoj svesci vršimo prvi pregled sirovih podataka. Cilj je da se upoznamo sa
strukturom skupa, proverimo tipove podataka i uočimo eventualne probleme
(nedostajuće vrednosti, duplikate, nelogične vrednosti) **pre** bilo kakve obrade.

**Važno:** u ovoj fazi podatke samo *posmatramo i dijagnostikujemo* — ne menjamo
ih i ne čistimo. Čišćenje i obrada slede u kasnijim sveskama, nakon podele na
trening i test skup.

## 1. Učitavanje podataka i biblioteka

Za rad koristimo biblioteku **pandas**, standardni alat za rad sa tabelarnim
podacima u Python-u. Učitavamo skup podataka iz CSV fajla u strukturu koja se
zove **DataFrame** (tabela sa redovima i kolonama).

In [1]:
# Uvozimo biblioteke potrebne za ovu fazu.
# pandas (pd) - za rad sa tabelama.
# numpy (np) - za numeričke operacije; koristićemo ga kasnije za rad sa NaN vrednostima.
import pandas as pd
import numpy as np

# Podešavanje da pandas prikazuje SVE kolone kada ispisujemo tabelu.
# Bez ovoga, pandas skraćuje prikaz i sakriva srednje kolone sa "...".
pd.set_option("display.max_columns", None)

# Učitavamo CSV fajl u DataFrame df.
# "../" znači da izlazimo iz foldera notebooks/ u glavni folder repoa.
df = pd.read_csv("../WA_Fn-UseC_-Telco-Customer-Churn.csv")

print("Podaci su uspešno učitani.")

Podaci su uspešno učitani.


## 2. Dimenzije skupa podataka

Prvo proveravamo osnovne dimenzije — koliko redova (korisnika) i koliko kolona
(atributa) skup sadrži.

In [2]:
# df.shape vraća par (broj_redova, broj_kolona).
print("Broj redova (korisnika):", df.shape[0])
print("Broj kolona (atributa):", df.shape[1])

Broj redova (korisnika): 7043
Broj kolona (atributa): 21


## 3. Prvi pogled na podatke

Prikazujemo prvih nekoliko redova kako bismo videli konkretne vrednosti i stekli
osećaj za sadržaj svake kolone.

In [3]:
# df.head() prikazuje prvih 5 redova tabele.
# Kao poslednja linija u ćeliji, Jupyter je automatski prikazuje kao tabelu.
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 4. Informacije o kolonama i tipovima podataka

Funkcija `info()` daje pregled svih kolona: njihov naziv, broj popunjenih
(ne-null) vrednosti i **tip podataka** (`dtype`).

Ovo je ključno za rano otkrivanje problema. Posebno nas zanima da li se tipovi
poklapaju sa očekivanjima — na primer, `TotalCharges` bi trebalo da bude broj,
pa ako je prepoznata kao tekst (`object`), to je znak da nešto nije u redu.

In [4]:
# df.info() ispisuje pregled: naziv kolone, broj ne-null vrednosti i tip podataka.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


## 5. Provera nedostajućih vrednosti

Standardnu proveru nedostajućih vrednosti radimo funkcijom `isnull()`, koja za
svaku ćeliju vraća `True` ako je vrednost nedostajuća (`NaN`), a `False` inače.
Sabiranjem tih vrednosti po kolonama dobijamo broj nedostajućih vrednosti u svakoj koloni.

In [5]:
# df.isnull() vraća tabelu True/False (True = vrednost nedostaje).
# .sum() sabira po kolonama: koliko True (nedostajućih) ima u svakoj koloni.
print("Broj nedostajućih vrednosti po kolonama:")
print(df.isnull().sum())

Broj nedostajućih vrednosti po kolonama:
customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


## 6. Skrivene nedostajuće vrednosti

Prethodna provera pokazala je da nema nedostajućih vrednosti — ali to može biti
obmanjujuće. U realnim podacima nedostajuće vrednosti se ponekad kriju kao
**prazni razmaci** (`" "`) umesto pravih `NaN` oznaka. Funkcija `isnull()` ih
ne prepoznaje jer su tehnički tekst, a ne prazne vrednosti.

Zato dodatno proveravamo da li neka kolona sadrži vrednosti koje su samo prazan razmak.

In [8]:
# Poredimo svaku ćeliju sa praznim razmakom " ".
# (df == " ") vraća tabelu True/False (True = ćelija sadrži tačno jedan razmak).
# .sum() sabira po kolonama koliko takvih "praznih" vrednosti ima.
print("Broj vrednosti koje su prazan razmak, po kolonama:")
print((df == " ").sum())

Broj vrednosti koje su prazan razmak, po kolonama:
customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


## 7. Analiza problema sa kolonom `TotalCharges`

Sada razumemo zašto je `TotalCharges` prepoznata kao tekst (`object`) umesto broja:
pošto 11 vrednosti sadrži prazan razmak (koji nije broj), pandas je celu kolonu
protumačio kao tekstualnu.

Da bismo mogli da analiziramo ovu kolonu numerički, moramo je konvertovati u
broj. Vrednosti koje se ne mogu konvertovati (prazni razmaci) pretvorićemo u
prave `NaN` oznake, koje potom možemo pravilno da analiziramo.

In [9]:
# pd.to_numeric() pokušava da pretvori kolonu u broj.
# errors="coerce" znači: vrednosti koje se NE mogu pretvoriti u broj
# (prazni razmaci) postaju NaN umesto da izazovu grešku.
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

# Ponovo proveravamo nedostajuće vrednosti — sada bi TotalCharges trebalo da
# pokaže 11 NaN vrednosti (bivši prazni razmaci).
print("Broj nedostajućih vrednosti nakon konverzije TotalCharges:")
print(df.isnull().sum())

Broj nedostajućih vrednosti nakon konverzije TotalCharges:
customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64


**Ovde zakljucujemo da TotalCharges bi trebalo da bude `numericka vrednost` sto nam govori da je treba da je konvertujemo**. To mozemo uraditi preko ugradjene funkcije u biblioteci `pandas`. Vrednosti koje ne mogu da se konvertuju postaju `NaN` vrednosti.

In [10]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"],errors="coerce")

## 8. Provera nedostajucih vrednosti nakon `konverzije` tipa

In [11]:
print("Nedostajuće vrednosti nakon konverzije TotalCharges:")
print(df.isnull().sum())
print()


Nedostajuće vrednosti nakon konverzije TotalCharges:
customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64



Ovde dolazimo do zakljucka da je `postoje` nedostajucih vrednosti u nasem skupu podataka i u sledecem koraku da utvrdimo koje su to vrednosti

In [13]:
print("Korisnici sa nedostajućim TotalCharges:")
print(df[df["TotalCharges"].isnull()])
print()

Korisnici sa nedostajućim TotalCharges:
      customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
488            No  No phone service        

## 9. Brisanje observacija 
Posto postoji samo `11` observacija nad nasim skupom podataka koji ima `7043 observacija` , dolazimo do zakljucka da mozemo samo `ukloniti` te observacije

In [ ]:
df = df.dropna(subset=["TotalCharges"])

7032


## 10. Brisanje duplikata
Sledece sto bi trebali proveriti jeste da li postoje `duplirani redovi` u nasem skupu podataka

In [25]:
print("Broj dupliranih redova:", df.duplicated().sum())

print("Broj dupliranih customerID vrednosti:",df["customerID"].duplicated().sum())
print()


Broj dupliranih redova: 0
Broj dupliranih customerID vrednosti: 0



Zakljucujemo da `nema duplikata` u nasem skupu podataka

## `Provera nakon ciscenja podataka`
Nakon ciscenja podataka treba da proverimo trenutno stanje naseg skupa podataka i da li je doslo do promena

## 1. Provera dimenzija
**Prvo treba da proverimo da li nam je broj observacija ostao isti**

In [ ]:

print("Broj redova nakon čišćenja:", df.dropna(subset=["TotalCharges"]).shape[0])
print("Broj kolona:", df.shape[1])
print()

Broj redova nakon čišćenja: 7032
Broj kolona: 21



## 2. Tipovi podataka
**Sledece sto bi trebali da uradimo jeste provera tipova podataka**

In [21]:
# Proveravamo tipove podataka nakon izvršenog čišćenja.
print("Tipovi podataka nakon čišćenja:")
print(df.dtypes)
print()

Tipovi podataka nakon čišćenja:
customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges        float64
Churn                object
dtype: object



## 3. Provera nedostajucih vrednosti
Takodje , trebali bi da proverimo da li smo zaista izbacili sve `NaN` vrednosti

In [22]:
print("Nedostajuće vrednosti nakon čišćenja:")
print(df.isnull().sum())
print()

Nedostajuće vrednosti nakon čišćenja:
customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64



### `Prikazivanje osnovne statistike za nas skup podataka`

In [23]:
print("Osnovna statistika numeričkih promenljivih:")
print(df[["SeniorCitizen", "tenure","MonthlyCharges", "TotalCharges"]].describe())
print()

Osnovna statistika numeričkih promenljivih:
       SeniorCitizen       tenure  MonthlyCharges  TotalCharges
count    7032.000000  7032.000000     7032.000000   7032.000000
mean        0.162400    32.421786       64.798208   2283.300441
std         0.368844    24.545260       30.085974   2266.771362
min         0.000000     1.000000       18.250000     18.800000
25%         0.000000     9.000000       35.587500    401.450000
50%         0.000000    29.000000       70.350000   1397.475000
75%         0.000000    55.000000       89.862500   3794.737500
max         1.000000    72.000000      118.750000   8684.800000



### Proveravamo koje vrednosti postoje u SeniorCitizen i koliko korisnika pripada svakoj kategoriji

In [26]:
print("Raspodela SeniorCitizen:")
print(df["SeniorCitizen"].value_counts())
print()

Raspodela SeniorCitizen:
SeniorCitizen
0    5890
1    1142
Name: count, dtype: int64



Izdvajamo sve kolone koje imaju `tekstualne` vrednosti

In [28]:
categorical_columns = df.select_dtypes(include="object").columns
print(categorical_columns)


Index(['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService',
       'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup',
       'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
       'Contract', 'PaperlessBilling', 'PaymentMethod', 'Churn'],
      dtype='object')


Za svaku kategoriju prikazujemo sve *`postojece`* kategorije i *`broj`* korisnika u svakoj kategoriji

In [29]:
for column in categorical_columns:
    print(f"\n{column}:")
    print(df[column].value_counts())


customerID:
customerID
7590-VHVEG    1
5575-GNVDE    1
3668-QPYBK    1
7795-CFOCW    1
9237-HQITU    1
             ..
6840-RESVB    1
2234-XADUH    1
4801-JZAZL    1
8361-LTMKD    1
3186-AJIEK    1
Name: count, Length: 7032, dtype: int64

gender:
gender
Male      3549
Female    3483
Name: count, dtype: int64

Partner:
Partner
No     3639
Yes    3393
Name: count, dtype: int64

Dependents:
Dependents
No     4933
Yes    2099
Name: count, dtype: int64

PhoneService:
PhoneService
Yes    6352
No      680
Name: count, dtype: int64

MultipleLines:
MultipleLines
No                  3385
Yes                 2967
No phone service     680
Name: count, dtype: int64

InternetService:
InternetService
Fiber optic    3096
DSL            2416
No             1520
Name: count, dtype: int64

OnlineSecurity:
OnlineSecurity
No                     3497
Yes                    2015
No internet service    1520
Name: count, dtype: int64

OnlineBackup:
OnlineBackup
No                     3087
Yes                